### ISSUE TITLE
python - swig_runtime_data conflict between builtin and non-builtin modules

### ISSUE THREAD

**Issue by eltoder** [swig/swig#1684]
It appears that SWIG types built with -builtin and without -builtin share the same swig_runtime_data capsule, because the flag is not encoded in the name. However, the types are not compatible. For example, when using -builtin, SwigPyObject is defined as

```
typedef struct {
  PyObject_HEAD
  void *ptr;
  swig_type_info *ty;
  int own;
  PyObject *next;
#ifdef SWIGPYTHON_BUILTIN
  PyObject *dict;
#endif
} SwigPyObject;
```

and tp_dealloc starts with

```
    SwigPyObject *sobj;                         \
    sobj = (SwigPyObject *)a;                   \
    Py_XDECREF(sobj->dict);                     \
```

If a module using swig without -builtin uses the same type, its type info can be overwritten with this one, which will crash on the Py_XDECREF line.

I suspect there are more incompatibilities between the two modes, so they should use separate capsules.

**Comment by vadz** [comment]
This does look like a problem, thanks. I'm not sure how do you end up with the same types in "builtin" and not "builtin" versions, but it would be better to fix this, of course.

AFAICS it should be as simple as adding `#ifdef SWIGPYTHON_BUILTIN/#else/#endif` around `SWIGPY_CAPSULE_NAME` definition in `Lib/python/pyhead.svg`, but I didn't test it and don't have a test case for it, so it would be nice if you could please make the change, check that it works and make a PR if it does. TIA!

**Comment by eltoder** [comment]
In our case, it was an instantiation of std::vector<double> that was used both by our module and cvxpy installed from pipy.

Yes, I think it should be simple. I'll try to make a PR.

**Comment by ojwb** [comment]
@eltoder Did you create a patch?  I can't seem to see a PR from you.

**Comment by ojwb** [comment]
Thanks for the quick response.

I certainly think something should be done about this - it's bad that two different Python modules created with SWIG can conflict behind the scenes like this just because one was built with `-builtin`.

I wasn't aware of `SWIG_TYPE_TABLE` before, but looking at it I wonder if maybe that should be set by default (to the module name say).  The case where you're using SWIG to create multiple modules which you actually want to share type information is the uncommon one, so having to add `-DSWIG_TYPE_TABLE=mymodulecollection` or equivalent when you actually want to share type information doesn't seem an unreasonable approach.  That also minimises exposure to the thread-unsafe aspects of the global type table (the manual warns about some of these, but people reasonably regard them as bugs - e.g. #103)

It looks like this would also be an efficiency gain when there are many SWIG-wrapped modules in use in a single Python program as the SWIG type handling seems to do a linear scan through all the known types for some operations (e.g. `SWIG_TypeQueryModule` when the type wasn't found by the name field).

@wsfulton Thoughts?

**Comment by eltoder** [comment]
@ojwb I made a PR: https://github.com/swig/swig/pull/2238

This is basically what @vadz said above, but also includes a unit test.

**Comment by wsfulton** [comment]
Agreed, Python with and without builtin are quite different and should be separate capsules. Fixed by merging in #2238.

As for changing the default from what is documented, gosh, I dunno, that'll be a brave move. Will probably break the universe where modules are relying on sharing type info, I'd be inclined to leave it as is. Maybe elaborate more in the docs in this area?
